### DISCUSSION TITLE
Image magick needs to generate previews as like Adobe Photoshop

### DISCUSSION THREAD

**Discussion by gippili** [ImageMagick/ImageMagick#6852] [Help]
### ImageMagick version

7.0.3-8

### Operating system, version and so on

Windows server 2016

### Description

Hi Team,
 
We have an expectation while uploading DNG files into our application. Our application uses ImageMagick to generate smaller size of the file for preview purposes.

When we upload raw DNG or adjusted (Modified in Adobe) DNG files in our application, we are expecting the preview generated by ImageMagick look same as the one  generated through Adobe Photoshop.

But current behavior is, previews generated by ImageMagick are different to what was generated in Adobe.
 
ImageMagick Command we are using:
 
magick.exe  DNG:Source.DNG -intent relative -sample 1024x768 -quality 95 -colorspace sRGB -strip -auto-orient -density 72 Destination.JPG
 
Modified command to match with Adobe setting:
magick.exe Source.DNG -intent relative -resize 1024x768 -quality 95 -colorspace Gray -gamma 1.4 -strip -auto-orient -density 72 Destination.JPG (here we changed our setting to have Gray gamma setting to match with Adobe, but each time this setting is different, so we can't change it every time.)
 
By Adobe, we got some gamma settings, so we changed the image magick command (version: 7.1.0-49) by adding gray gamma settings then we got result which is close to Adobe generated preview.
 
Here the challenge is, we are not sure what are the settings being done by Adobe. So we are expecting ImageMagick to generate same type of results without changing the magick command.
 
ImageMagick needs to be generating the previews based on the settings done by Adobe automatically. Without changing any settings on our application's end.
 
If there's any such way, please let us know.
 
Please suggest a way to achieve the desired results which should similar to Adobe generated previews.
  
Thanks
Gopal

**Comment by snibgo** [comment]
Photoshop is showing either the JPG made by the camera that is embedded in the raw file, or it is processing the raw data to "prettify" it.

You can extract the embedded JPG with exiftool.

Adobe don't publish their prettifying algorithm. It is probably complex, and would include a tonal shift that may be similar to IM's "`-auto-gamma`", and sharpening.

**Comment by gippili** [comment]
Hi Team,

Thank you for the response!

We want to check with you again on this issue more precisely.
We are expecting ImageMagick to generate previews as like adobe photoshop but not sure what are settings used for each adjusted dng files in Adobe.
Is there a way to achieve desired results by integrating any new packages from ImageMagick like Adobe or any other tool? Is there anything available like that in newer versions of ImageMagick where it supports and generates same images as it's in Adobe.
We are not sure what properties/settings adjusted DNG files have from Adobe, so it's difficult to have each ImageMagick command for each Adobe setting. We want to know if there's any option like , not changing the ImageMagick commands for each type of setting from Adobe and generate renditions just like how it's done in Adobe.
We are just looking for a solution without doing any customization to our application.

Regards
Gopal

**Comment by snibgo** [comment]
I suggest you zip your DNG file, and the preview that Adobe Photoshop has made, and post that zip file here. I may be able to suggest a suitable command for you.

**Comment by gippili** [comment]
Hi,

As requested, I was trying to share but the adjusted dng files I have those flies have big size(greater than 27MB but here allowing 25MB only). Even we tried to compress over the internet but somehow not able to compress.
Is there any other way to share that dng file?

Please suggest.

Thanks
Gopal

**Comment by snibgo** [comment]
Put it anywhere on the internet, and paste the URL here.

**Comment by gippili** [comment]
As requested, I am sharing those files through google drive. Here's the link:
https://drive.google.com/file/d/1kbG9qGgP5hkqA7zC2cbGjycak3fq5ATB/view?usp=drive_link
username: gopi.sharefile@gmail.com
passwrd: Share!12345

NOTE: I have created this account for this purpose. Plz let me know after download of zip file so that I can delete this account.

1) Adjusted dng file, the one we are uploading in our application.
2) Since we don't have Adobe Photoshop in our local, we are sharing the screenshot which was shared by customer.
3) Later we have applied gray gamma setting to the image magick command then we got the result which is close to adobe generated preview. If we apply this globally, other dng files also(raw dng) got effected(means changed to black & white) but which are not supposed to. Here's the command:

    magick.exe Source.DNG -intent relative -resize 1024x768 -quality 95 -colorspace Gray -gamma 1.4 -strip -auto-orient -density 
    72 Destination1.JPG

As discussed, we are not sure what setting(there are n number of settings like gray-gamma) they use to adjust the image but when we upload, have to identify the setting automatically & get the desired preview similar to adobe generated preview. If we can't achieve through image magick alone, which third party tool we can integrate to get the desired results.

yeh, I am looking forward to your recommendations.

Thanks
Gopal

**Comment by snibgo** [comment]
In the screenshot, the "Saturation" slider is all the way down, at "-100". So the resulting image is grayscale. In ImageMagick, use "`-colorspace Gray`" for this.

`-auto-level` ensures the image spans the entire gray range from 0 to 100%.

`-auto-gamma` applies a power curve to put the mean level at about 50%.

`-sigmoidal-contrast 2,50%` increases the contrast in middle tones (near 50%), while decreasing contrast at extreme shadows and highlights.

```
magick Adjusted_DNG.DNG -colorspace Gray -auto-level -auto-gamma -sigmoidal-contrast 2,50% x2.tiff
```
The result, `x2.tiff`, is very similar to the screenshot.

**Reply by fmw42** [reply]
You can also use -modulate 100,0,100 to emulate the saturation slider in at -100 in Photoshop in place of -colorspace gray.  There will be slight differences between -modulate and -colorspace results. You might get closer to your Photoshop result using -modulate.

**Comment by gippili** [comment]
Thanks for the information!!

As suggested, we tried below command for Adjusted DNG file & getting preview close to adobe generated preview. Attaching screenshots here for reference.
magick Adjusted_DNG.DNG -colorspace Gray -auto-level -auto-gamma -sigmoidal-contrast 2,50% DestinationPreview.JPG

Also same we played around the command with -modulate for Adjusted DNG file, here also we got similar result as like preview which were generated through above command. 
magick Adjusted_DNG.DNG -modulate 100,0,100 -auto-level -auto-gamma -sigmoidal-contrast 2,50% DestinationPreview.JPG

But the challenge here is, if we apply this command even for raw dng file which was not adjusted with any setting, the generated preview should be similar to original file, not black & white(which was not expected, attached same).

All in all, we are looking for a solution in such a way that, single image magick command should identify whether the uploaded dng file is adjusted with any setting like gray-gamma or its a raw one. For instance, If we upload adjusted dng having x setting, the preview should generate automatically or if we upload raw camera dng file, the preview should similar to raw dng.

If its not achievable by image magick single command, what other tools we can integrate to make this work as expected.

Your recommendations are really valuable for us.

Regards, Gopal
![Uploading Orginal&generatedpreviews.png…]()

**Comment by gippili** [comment]
Refer attachment here, if you guys not able to see there
[Orginal&generatedpreviews.zip](https://github.com/ImageMagick/ImageMagick/files/13378275/Orginal.generatedpreviews.zip)

**Comment by snibgo** [comment]
> ... the generated preview should be similar to original file, not black & white ...

You can remove the part of the command that makes the result black and white, so remove `-colorspace Gray` or `-modulate 100,0,100`.

> If we upload adjusted dng having x setting, the preview should generate automatically or if we upload raw camera dng file, the preview should similar to raw dng.

How does any software know the dng is adjusted with "x setting"?
